zerver/lib/bot_lib.py

```diff
@@ -32,6 +34,58 @@ def get_bot_handler(service_name):
     bot_module = importlib.import_module(bot_module_name)  # type: Any
     return bot_module.handler_class()
 
+class StateHandler(object):
+    # Cache the state size of each bot so we don't have to do expensive accumulations
+    # each time the bot receives a message.
+    state_sizes = {}  # type: Dict[int, int]
+    state_size_limit = 10000000   # type: int # TODO: Store this in the server configuration model.
+
+    def __init__(self, user_profile):
+        # type: (UserProfile) -> None
+        self.user_profile = user_profile
+        self.marshal = lambda obj: obj
+        self.demarshal = lambda obj: obj
+
+    @contextmanager
+    def __call__(self, marshal=None, demarshal=None):
+        # type: (Optional[Callable[[Any], Text]], Optional[Callable[[Any], Text]]) -> Iterator[StateHandler]
+        prev_marshal = self.marshal
+        prev_demarshal = self.demarshal
+        self.marshal = marshal or self.marshal
+        self.demarshal = demarshal or self.demarshal
+        yield self
+        self.marshal = prev_marshal
+        self.demarshal = prev_demarshal
```

Comment:
You should use a consistent number of `l`s for "marshall".  

Also, can you explain why you need this code with the yield etc.?  I wouldn't expect it to be required, since I'd imagine we'd just pass these fields to a class instance and they'd be consistently correct for that instance.

Reply:
I found the one `l` version the most common online  - should I still use "marshall" ?

Reply:
I used that code with the yield because I'd expect a context manager to always reset the contextualized object to the original state.  But I see that it'd probably be easier to just make the cm argumentless and set any field prior to the call.

Reply:
yeah, `marshal` seems more common.

I think it just isn't clear to me why we need to restore state in the `__init__` method of the class; the "restored original" state won't ever be used, will it?  But maybe I'm missing something.

Reply:
I updated the PR to have the marshal and demarshal attributes be set as class attributes, instead of in the context manager, if that's what you mean.